CODE: genetic-value

daikitag opened the pull request:
Create genetic_value function that gives a deterministic result based on the trait dataframe.

jeromekelleher commented on tstrait/genetic.py:
```diff
@@ -0,0 +1,207 @@
+        self.trait_df = trait_df[["site_id", "effect_size", "trait_id", "causal_state"]]
+        self.ts = ts
+
+    def _individual_genotype(self, tree, site, causal_state, num_nodes):
```
There's isn't really much point in counting the causal states, and we may as well just accumulate the actual values here and save the multiply later. Something like
```
def _individual_genetic_values(self, tree, site, causal_state, effect_size):
     has_mutation = np.zeros(self.ts.num_nodes + 1, dtype=bool)

jeromekelleher commented on tstrait/genetic.py:
```diff
@@ -0,0 +1,207 @@
+
+
+@numba.njit
+def _traversal_genotype(
```
As below, ``genotype`` isn't really the right word and we may as well sum the effect sizes directly.

jeromekelleher commented on tstrait/genetic.py:
```diff
@@ -0,0 +1,207 @@
+        else:
+            individual_id = nodes_individual[parent_node_id]
+        if individual_id > -1:
+            genotype[individual_id] += 1
```
genetic_value[individual] += effect_size

jeromekelleher commented on tstrait/genetic.py:
```diff
@@ -0,0 +1,207 @@
+    -------
+    pandas.DataFrame
+        Pandas dataframe that includes genetic value of individuals in the
+        tree sequence data.
```
"tree sequence data" isn't necessary - just say "tree sequence."

jeromekelleher commented on tstrait/genetic.py:
```diff
@@ -0,0 +1,207 @@
+
+        * **site_id**: Site IDs that have causal mutation.
+        * **effect_size**: Simulated effect size of causal mutation.
+        * **causal_state**: Derived state of causal mutation.
```
This isn't correct - we can have multiple causal mutations. Our model is a causal *allele*, right?

daikitag replied:
Just to be clear, am I correct with saying that we should name the `causal_state` column to be `causal_allele` instead?

jeromekelleher replied:
Yes, that's what we're discussing in #113 

I'm pointing out that the documentation is incorrect here though, because you say causal *mutation*. We can have more than one mutation to the causal allele.

jeromekelleher commented on tstrait/genetic.py:
```diff
@@ -0,0 +1,207 @@
+
+        * Trait IDs in **trait_id** column must start from zero and be consecutive.
+
+    The phenotype output of the simulation is given as a :py:class:`pandas.DataFrame`.
```
This line shouldn't be here.

jeromekelleher reviewed (commented):
LGTM, minor comments. We need to be clearer about the distinction between alleles and mutations, though.

jeromekelleher commented:
Let's finish the ``sim_trait`` function fully before trying to get this done any further.

daikitag commented:
@jeromekelleher This is completely finished now.
Would it be possible for you to check it whenever you have some time?
I just modified the `genetic_value` function, and I also modified the documentation.
There are other functions being modified, but they are just modified to account for the fact that the naming of the functions are different now.

daikitag commented:
This should be seen after https://github.com/tskit-dev/tstrait/pull/115

jeromekelleher commented:
I'll have a look at this when the trait functions etc have been merged.

daikitag commented:
> I'll have a look at this when the trait functions etc have been merged.

Thanks Jerome. I'm sorry that the changes being made are really big in this pull request, as I had to change all files and documentation to pass the tests.
